test/suites/openfga.sh

```diff
@@ -0,0 +1,282 @@
+  user_is_not_server_admin
+  user_is_server_operator
+  user_is_project_manager
+  user_is_project_operator
+
+  # Give the user the `manager` entitlement on `project:default`.
+  fga tuple delete --store-id "${OPENFGA_STORE_ID}" user:user1 operator server:lxd
+  fga tuple write --store-id "${OPENFGA_STORE_ID}" user:user1 manager project:default
+
+  echo "==> Checking permissions for project manager..."
+  user_is_not_server_admin
+  user_is_not_server_operator
+  user_is_project_manager
+  user_is_project_operator
+
+  # Give the user the `operator` entitlement on `project:default`.
+  fga tuple delete --store-id "${OPENFGA_STORE_ID}" user:user1 manager project:default
+  fga tuple write --store-id "${OPENFGA_STORE_ID}" user:user1 operator project:default
+
+  echo "==> Checking permissions for project operator..."
+  user_is_not_server_admin
+  user_is_not_server_operator
+  user_is_not_project_manager
+  user_is_project_operator
+
+  # Change permission to "user" for instance "foo"
+  lxc launch testimage foo
+  fga tuple delete --store-id "${OPENFGA_STORE_ID}" user:user1 operator project:default
+  fga tuple write --store-id "${OPENFGA_STORE_ID}" user:user1 user instance:default/foo
+
+  echo "==> Checking permissions for instance user..."
+  user_is_instance_user foo
+  lxc delete foo --force
+  user_is_not_server_admin
+  user_is_not_server_operator
+  user_is_not_project_manager
+  user_is_not_project_operator
+
+  # Unset config keys.
+  lxc config unset candid.api.url
+  lxc config unset candid.api.key
+  lxc config unset openfga.api.url
+  lxc config unset openfga.api.token
+  lxc config unset openfga.store.id
+  lxc remote rm candid-openfga
+
+  shutdown_openfga
+}
+
+user_is_not_server_admin() {
+  # Can always see server info (type-bound public access https://openfga.dev/docs/modeling/public-access).
+  lxc_remote info candid-openfga: > /dev/null
+
+  # Cannot see any config.
+  ! lxc_remote info candid-openfga: | grep -Fq 'core.https_address' || false
+
+  # Cannot set any config.
+  ! lxc_remote config set candid-openfga: core.proxy_https=https://example.com || false
+
+  # Should still be able to list storage pools but not be able to see any storage pool config or delete.
+  [ "$(lxc_remote storage list candid-openfga: -f csv | wc -l)" = 1 ]
+  lxc storage create test-pool dir
+  ! lxc_remote storage set candid-openfga:test-pool rsync.compression=true || false
+  ! lxc_remote storage show candid-openfga:test-pool | grep -Fq 'source:' || false
+  ! lxc_remote storage delete candid-openfga:test-pool || false
+  lxc storage delete test-pool
+
+  # Should not be able to create a storage pool.
+  ! lxc_remote storage create candid-openfga:test dir || false
+
+  # Should still be able to list certificates
+  [ "$(lxc_remote config trust list candid-openfga: -f csv | wc -l)" = 1 ]
+
+  fingerprint="$(lxc config trust list -f csv | cut -d, -f4)"
+
+  # But cannot edit non-certificate fields.
```

Comment:
Do you mean although the user is not an admin, they can still change their own certificate? If so, doing that via yaml will be pretty gnarly... I could at least double check that they are able to revoke their own certificate though.

Reply:
Ah of course, this no longer makes sense because we're using candid authentication. The certificate we're deleting isn't "ours" in the sense that it's what is used to authenticate when we use the `localhost` remote, not the `candid-openfga` remote.